Ditch DirectX in favor of OpenGL; use SDL window; ditch Windows Registry

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [ ] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

As a reminder:

- (NEW) F1/Shift-F1: toggle upscaling 1-2-3-4× ("scaler")
- F2/Shift-F2: increase or decrease viewport size ("sizer")
- F3: toggle bilinear filter
- F4: toggle perspective filter
- F7: toggle zbuffer
- F12: toggle SWR/HWR renderer

Known issues:

- [x] There are black rectangles around the window on some configurations
- [x] The game sometimes crashes when using the new scaler option (F1 / Shift-F1)
- [x] Scaler/sizer woes with HWR/SWR
- [x] Console fails to get drawn on non-1x-scaling
- [ ] Window isn't brought to front on launch
- [ ] Text doesn't respect aspect ratio
- [x] Aspect ratio issues
- [x] ZBuffer near plane clipping (looking closely at door objects, passport animation)
- [x] Alpha problems for UI in HWR
- [x] Alpha problems for UI in SWR
- [x] Alpha edge problems in textures in HWR
- [x] Ingame backgrounds do not work at all (neither SWR nor HWR)
- [x] 3D pickups overlay clips with HWR + ZBuffer ON
- [x] The controls UI doesn't display
- [x] The ammo counter doesn't display
- [x] Health/air bars don't show in HWR + ZBuffer ON
- [x] Health/air bars appear ugly in HWR
- [x] The health bar doesn't display when highlighting a medi pack in the inventory
- [x] If you use F5/F6 or flip through the passport pages - the OPTION text blinks. It's subtle but noticeable.
- [x] Green pixels appear in explosion sprites in Dragon's Lair (level file textures seem OK, doesn't happen in OG)
- [x] Fade out in the title screen clips the inventory items in HWR / ZBuffer ON
- [x] The end of level stats background uses the main menu image instead, with these steps (`/play 1`, `/endlevel`)
- [x] The end of level stats background looks bugged (`/play 1`, open then close the inventory, `/endlevel`, skip cutscene)
- [x] The textures keep zooming in when toggling the renderer with the bilinear filter on
- [x] Quitting the game with Alt+F4 goes to the inventory very briefly(?)
- [x]  The end of level stats background is now a black screen. In OG, the end of level stats uses the carpet background like the inventory.
- [x] The game crashes with these steps: `/play 1`, open the inventory or F5 or F6, `/title`
- [x] The game crashes with these steps: `/play 1`, open the inventory or F5 or F6, `/demo`, ending the demo crashes the game
- [x]  If TR2X is behind another window e.g. notepad, while in the main menu, the game freezes and I get a black screen when the demo mode starts. Conversely, this issue happens when the demo ends while another window is in front of TR2X
- [x] In the inventory, when the medi pack is highlighted and you switch to another item, the health bar persists until the next item is highlighted. In OG, the health bar would disappear the moment you press left/right to switch.
- [x] Similar issue to above, if you have a medi pack highlighted when switching to rings above or below, the health bar is slow to disappear and quick to reappear. In OG, the health bar would appear/disappear at the same time as the green arrows in the middle ring.
- [x] The water/acid/mercury/lava animated textures is now static.
- [x] SWR crash related to bad default UVs
- [x] Using F1/Shift+F1 while in the stopwatch/end of level stats causes the UI to become persistent
- [x] Looking around in the Floater crashes the game when trying to draw too many rooms at once

walkawayy commented:
Honestly this plays really nicely! Credit images fade in / fade out seems to work fine.

Some issues I noticed (at least with HWR):

- No inventory red waving background. Just a black screen.
- 3D pickups in the overlay are sometimes behind geometry in the game world like the floor. They also don't clip their meshes properly.
- The Bartoli dragon coming alive might need checked? @aredfan I'm not super familiar with it's OG behavior.

lahm86 commented:
Looking great so far! Other than the issues mentioned, I've noticed the following.

- Window isn't brought to front on launch
- Health/air bars don't show in HWR
- Full screen aspect ratio issue maybe? 2560x1440 suffers from "wide/squashed Lara" issue; maximized window-mode doesn't have this problem.
![image](https://github.com/user-attachments/assets/eec91221-082f-4e41-935c-2b78783963b8)

- Green pixels appear in explosion sprites in Dragon's Lair (level file textures seem OK, doesn't happen in OG)
![image](https://github.com/user-attachments/assets/eafbcb0a-f99c-48c7-8931-53f3356f3f98)

aredfan commented:
I made a note of these few issues after a short time testing on HWR.

- [x] The controls UI doesn't display.
- [x] The ammo counter doesn't display.
- [x] The colour bitrate is lower in HWR compared to before, maybe it's using 24 instead of 32?

@walkawayy I didn't notice anything out of the ordinary with the dragon but I can check again soon.

rr- commented:
> * [ ]  The colour bitrate is lower in HWR compared to before, maybe it's using 24 instead of 32?

Can I get direct comparison screenshots? Thanks. If anything it's using 8-bit palette rather than the 16-bit one, but I didn't notice differences so I'd appreciate a more direct example.

aredfan commented:
> > * [ ]   The colour bitrate is lower in HWR compared to before, maybe it's using 24 instead of 32?
> 
> Can I get direct comparison screenshots? Thanks. If anything it's using 8-bit palette rather than the 16-bit one, but I didn't notice differences so I'd appreciate a more direct example.

Yes of course. See the colour bandings in the passport pages in HWR.

**PR:**
![20241203_125135_The_Dragons_Lair](https://github.com/user-attachments/assets/ffa5b6e6-e130-4f89-a4a9-e8c6f7dcd941)
**Dev snapshot:**
![20241203_125248_The_Dragons_Lair](https://github.com/user-attachments/assets/31b0f5d8-f0a1-45d8-8ab8-8b995987ef00)

The fire sprite has a mixture of orange and yellow in the dev snapshot, whereas it's mostly orange in the PR.
**PR:**
![20241203_125149_The_Dragons_Lair](https://github.com/user-attachments/assets/b4ba9ac2-0bbf-4b74-862b-6d9e57feff1d)
**Dev snapshot:**
![20241203_125252_The_Dragons_Lair](https://github.com/user-attachments/assets/a7576da7-3313-42df-bc04-a7c69c0d4ff3)

rr- commented:
@aredfan what if you use windowed mode and disable 16-bit textures?

aredfan commented:
The controls UI is fixed, thank you, but the transparent backgrounds are now more opaque.  More screenshots to follow.

Left is latest dev snapshot, right is PR.

**HWR:**
![1](https://github.com/user-attachments/assets/0533aebc-877a-46ba-b623-c1da47ceefea)
![2](https://github.com/user-attachments/assets/c0587bed-cd03-42a9-8cbb-d2fb900ae43c)
**SWR:**
![3](https://github.com/user-attachments/assets/8774c87a-6d82-4bcd-82af-bf361444269f)

**Edit:** I noticed the alpha textures appears different along the edges, when bilinear filter is enabled.

Left is latest dev snapshot, right is PR.
![1](https://github.com/user-attachments/assets/38768604-0018-4255-86cb-c66a8e837268)

rr- commented:
@lahm86 can you check the newest build (once it finishes building) for this?

>Window isn't brought to front on launch

I can't test this part reliably on wine.

lahm86 commented:
> @lahm86 can you check the newest build (once it finishes building) for this?
> 
> > Window isn't brought to front on launch
> 
> I can't test this part reliably on wine.

Fixed! Thank you.

aredfan commented:
The UI alpha issue in HWR is fully resolved, thank you. 👍 

There's UI alpha issues remaining in SWR. Whilst the customize controls section looks good. 👇 

![20241203_165435_Laras_Home](https://github.com/user-attachments/assets/08d25f24-e1c0-4d94-b02f-20cf5777285f)

Whereas other aspects of the main menu are showing black parts in the transparent backgrounds.

Left is latest dev snapshot, right is the PR build.
![1](https://github.com/user-attachments/assets/d3fa1b95-1218-4881-b894-93ac2d69d2c8)
![2](https://github.com/user-attachments/assets/765202c4-d544-4f5d-b08c-0fca6769d7d9)
![3](https://github.com/user-attachments/assets/55f0995c-26fb-4900-8913-181935539e0b)

_____________

**Edit:** New issue possibly. This only happens when z-buffer is enabled. Lara's Home, room 55.

Left is latest dev snapshot, right is the PR build.
![1](https://github.com/user-attachments/assets/3f29eaf8-1a29-4b9e-84c0-1b47ad2d7baa)

_______________

- [ ] The health bar doesn't display when highlighting a medi pack in the inventory.
- [ ] If  you use F5/F6 or flip through the passport pages - the OPTION text blinks. It's subtle but noticeable.

rr- commented:
@aredfan re: weird colors of UI, that's just the way it is with 8-bit palettes. Notice how the good part is actually rendered on top of the keys meshes which have a different background.

aredfan commented:
I found these issues while testing.

_________________

The end of level stats background uses the main menu image instead, with these steps:

1. `/play 1`
2. `/L`

![20241206_180202_The_Great_Wall](https://github.com/user-attachments/assets/271d319b-0bae-4a46-b914-55801d6c34cb)

_________________________

The end of level stats background looks bugged, with these steps:

1. `/play 1`
2. open then close the inventory
4. `/L`
5. skip cutscene

![20241206_180220_The_Great_Wall](https://github.com/user-attachments/assets/f1f1042f-7251-4d48-8247-78e0f20beb6a)

______________________

The game crashes with these steps:

1. `/play 1`
2. open then close the inventory
3. `/title`

or

1. `/play 4`
2. open then close the inventory
3. `/L`

__________________________

I found this new issue, if you have bilinear filter enabled and keep pressing F12, the textures will keep zooming in.

![20241206_182600_The_Great_Wall](https://github.com/user-attachments/assets/bad9c548-3d33-404c-80a7-9d6fe3556f9d)

lahm86 commented:
A few things noticed (clean config/install):
- Wide/squashed Lara now happens (for me at least) in either windowed of full-screen mode
- The issue of the window not coming to the front seems to have regressed
- The bars are cut off in windowed mode
![image](https://github.com/user-attachments/assets/1e2ae305-0bd8-48b8-be34-b4aeb97d9c43)
![image](https://github.com/user-attachments/assets/afc1a57d-48ea-4bbe-ad59-ec683081a3cc)

rr- commented:
>Wide/squashed Lara now happens (for me at least) in either windowed of full-screen mode

This is likely caused by the aspect ratio mode setting. ((Just a quick reminder, it's now located in the config instead of `-setup`, and it should be possible to change it at runtime with `/set`). I'll keep this tracked under the aspect ratio issues checkbox.

>The issue of the window not coming to the front seems to have regressed

Can I ask for help with this one? The original fix was to call `SDL_RaiseWindow` after `SDL_ShowWindow` which hasn't been touched since.

>The bars are cut off in windowed mode

I'll tackle this as part of the black rectangles issue.

aredfan commented:
The regressions in my previous reply are fixed, thank you.  There are some more issues I found while doing further tests.

___________________

- [x] The end of level stats background is now a black screen. In OG, the end of level stats uses the carpet background like the inventory.

_____________________

- [x] The game crashes with these steps:

1. `/play 1`
2. open the inventory or F5 or F6
3. `/title`

___________________

- [x] The game crashes with these steps:

1. `/play 1`
2. open the inventory or F5 or F6
3. `/demo`
4. ending the demo crashes the game

______________________

- [x] If TR2X is behind another window e.g. notepad, while in the main menu, the game freezes and I get a black screen when the demo mode starts. Conversely, this issue happens when the demo ends while another window is in front of TR2X

______________________

- [x] In the inventory, when the medi pack is highlighted and you switch to another item, the health bar persists until the next item is highlighted. In OG, the health bar would disappear the moment you press left/right to switch.
- [x] Similar issue to above, if you have a medi pack highlighted when switching to rings above or below, the health bar is slow to disappear and quick to reappear. In OG, the health bar would appear/disappear at the same time as the green arrows in the middle ring.

_____________________

- [x] The water/acid/mercury/lava animated textures is now static.

lahm86 commented:
> Wide/squashed Lara

Got it, apologies. Yes, it looks OK on 16:9 in windowed mode again (so just full-screen issue remains). Maybe this should be the default aspect ration as the default window size is 1280x720? Would it be helpful if I provide a diff for the config tool option additions? Or if you'd like to keep that as a separate PR, that's understandable.

> Cut-off bars in windowed mode

Just to add here that this only happens in windowed 4:3 aspect ratio, 16:9 appears normal.

> Window to front

I can't find the solution to this. I am wondering if it's Windows blocking it for some reason, there are a lot of criteria seemingly for setting the foreground window (I assume SDL is eventually calling this Windows function under the hood). Perhaps it's related to the fact the exe is still Tomb2.exe and Windows thinks we're trying to steal focus (although in my head the criteria is met so I'm really not sure). It's certainly not an issue in TR1X, so maybe we come back to address this one at a later date.
https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow#remarks

rr- commented:
> > Wide/squashed Lara
> 
> Got it, apologies. Yes, it looks OK on 16:9 in windowed mode again (so just full-screen issue remains). Maybe this should be the default aspect ration as the default window size is 1280x720? Would it be helpful if I provide a diff for the config tool option additions? Or if you'd like to keep that as a separate PR, that's understandable.

Yeah – let's default to 16:9. I also plan to add letterboxing in case the player chooses 4:3 aspect mode but has a bigger window. I suppose we should provide some degree of configuration in the config tool, but IMO not all settings make sense, such as window width or height, or the fullscreen option. In any event I'd appreciate a diff to consolidate into this PR with the rest of the settings.

> > Cut-off bars in windowed mode
> 
> Just to add here that this only happens in windowed 4:3 aspect ratio, 16:9 appears normal.

Thanks, but I believe in general the viewport logic needs adjustment as it's clearly cut when the window size is not a multiple of 32.

> > Window to front
> 
> I can't find the solution to this. I am wondering if it's Windows blocking it for some reason, there are a lot of criteria seemingly for setting the foreground window (I assume SDL is eventually calling this Windows function under the hood). Perhaps it's related to the fact the exe is still Tomb2.exe and Windows thinks we're trying to steal focus (although in my head the criteria is met so I'm really not sure). It's certainly not an issue in TR1X, so maybe we come back to address this one at a later date. https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow#remarks

Thanks for looking into it. Yeah, I think it's best to tackle it as a follow-up issue.

lahm86 commented:
@rr- LMK if this looks OK. I've only added three config options: aspect ratio, fade effects and mic at Lara. Everything else I agree doesn't make sense in the config tool.
I've updated `camera.c` too to reference the mic setting. I suppose we could do away with `g_Camera.is_lara_mic` but I'm not sure if it's used anywhere else in code yet to be decompiled. I've checked this and it works, so e.g. `/play 0 /tp 43 2 89` - depending on the setting, you will either hear the clock after pushing the button when the camera shifts or not.

```diff
diff --git a/src/tr2/config_map.def b/src/tr2/config_map.def
index 1f5f6202..e4dfa1ab 100644
--- a/src/tr2/config_map.def
+++ b/src/tr2/config_map.def
@@ -9,7 +9,7 @@ CFG_BOOL(g_Config, visuals.enable_fade_effects, true)
 CFG_BOOL(g_Config, visuals.fix_item_rots, true)
 CFG_ENUM(g_Config, rendering.screenshot_format, SCREENSHOT_FORMAT_JPEG, SCREENSHOT_FORMAT)
 CFG_ENUM(g_Config, rendering.render_mode, RM_HARDWARE, RENDER_MODE)
-CFG_ENUM(g_Config, rendering.aspect_mode, AM_ANY, ASPECT_MODE)
+CFG_ENUM(g_Config, rendering.aspect_mode, AM_16_9, ASPECT_MODE)
 CFG_ENUM(g_Config, rendering.texture_filter, GFX_TF_NN, GFX_TEXTURE_FILTER)
 CFG_BOOL(g_Config, rendering.enable_zbuffer, true)
 CFG_BOOL(g_Config, rendering.enable_perspective_filter, true)
diff --git a/src/tr2/game/camera.c b/src/tr2/game/camera.c
index 4da1e9e3..f2f8e13b 100644
--- a/src/tr2/game/camera.c
+++ b/src/tr2/game/camera.c
@@ -31,6 +31,7 @@
 
 void __cdecl Camera_Initialise(void)
 {
+    g_Camera.is_lara_mic = g_Config.audio.enable_lara_mic;
     Camera_ResetPosition();
     Output_AlterFOV(GAME_FOV * PHD_DEGREE);
     Camera_Update();
diff --git a/tools/tr2/config/TR2X_ConfigTool/Resources/Lang/en.json b/tools/tr2/config/TR2X_ConfigTool/Resources/Lang/en.json
index a567aa39..00060384 100644
--- a/tools/tr2/config/TR2X_ConfigTool/Resources/Lang/en.json
+++ b/tools/tr2/config/TR2X_ConfigTool/Resources/Lang/en.json
@@ -10,6 +10,11 @@
     "screenshot_format": {
       "jpg": "JPEG",
       "png": "PNG"
+    },
+    "aspect_mode": {
+      "any": "Any",
+      "16:9": "16:9",
+      "4:3": "4:3"
     }
   },
   "Properties": {
@@ -41,9 +46,21 @@
       "Title": "Key item pre-selection",
       "Description": "When Lara presses action against a keyhole or puzzle slot, and she has the corresponding item in the inventory, that item will be pre-selected."
     },
+    "aspect_mode": {
+      "Title": "Aspect mode",
+      "Description": "Set the aspect ratio of the game output."
+    },
     "screenshot_format": {
       "Title": "Screenshot format",
       "Description": "Screenshot file format."
+    },
+    "enable_lara_mic" : {
+      "Title": "Microphone at Lara",
+      "Description": "Set the microphone to be at Lara's position. If disabled, the microphone will be at the camera's position."
+    },
+    "enable_fade_effects": {
+      "Title": "Fade effects",
+      "Description": "Enable fade transitions, for example between credit graphics."
     }
   }
 }
diff --git a/tools/tr2/config/TR2X_ConfigTool/Resources/specification.json b/tools/tr2/config/TR2X_ConfigTool/Resources/specification.json
index 53c99167..e3a71e8b 100644
--- a/tools/tr2/config/TR2X_ConfigTool/Resources/specification.json
+++ b/tools/tr2/config/TR2X_ConfigTool/Resources/specification.json
@@ -3,6 +3,11 @@
     "screenshot_format": [
       "jpg",
       "png"
+    ],
+    "aspect_mode": [
+      "any",
+      "16:9",
+      "4:3"
     ]
   },
   "CategorisedProperties": [
@@ -70,6 +75,12 @@
           "DataType": "Bool",
           "DefaultValue": true
         },
+        {
+          "Field": "aspect_mode",
+          "DataType": "Enum",
+          "EnumKey": "aspect_mode",
+          "DefaultValue": "16:9"
+        },
         {
           "Field": "screenshot_format",
           "DataType": "Enum",
@@ -82,14 +93,22 @@
       "ID": "sound",
       "Image": "Graphics/graphic6.jpg",
       "Properties": [
-
+        {
+          "Field": "enable_lara_mic",
+          "DataType": "Bool",
+          "DefaultValue": false
+        }
       ]
     },
     {
       "ID": "ui",
       "Image": "Graphics/graphic7.jpg",
       "Properties": [
-
+        {
+          "Field": "enable_fade_effects",
+          "DataType": "Bool",
+          "DefaultValue": true
+        }
       ]
     }
   ]
```

rr- commented:
I believe the g_Camera fix should go as a separately tracked issue/PR pair as it seems to be an OG bug.

walkawayy commented:
I just tested the latest version with the HWR and:

- The game window was brought to the front for me with no issue. I had a window open, launched the game, and it appeared at the front with no issue.
- `Console fails to get drawn on non-1x-scaling` only happens to me in windowed mode. At scalings 2-4, the console is drawn fine for me in fullscreen. It looks like the font is getting bigger though in windowed mode at these scalings which I'm guessing means the console is getting drawn offscreen or something?
- I couldn't reproduce a crash using the new scaling option. I tried F1 / shift F1 ing a lot with no issue.
- I don't know what the F2 key does, but it seems broken. It moves the game window increasingly offscreen in both windowed and fullscreen. Here's a screenshot of what happens:
![image](https://github.com/user-attachments/assets/09240663-af84-4c90-aa4d-a1e1c8b40272)

lahm86 commented:
I've noticed since previous checks the skybox appearing in the distance in e.g. Offshore Rig, but I'm wondering if this can be fixed by introducing draw dist/fade settings. And if so, it'd be fine as a follow up task IMO. I don't think this is a new issue in the PR by the way, I just skipped past certain levels with the problem.

![image](https://github.com/user-attachments/assets/075d5cfc-0f4f-4a5c-826f-876f9b7abc20)

One bigger issue has appeared though since testing yesterday: SWR mode crashes for me in windowed mode, but it seems to be dependent on the window size. Unfortunately the log file doesn't produce a trace. The default window size is fine, but it consistently crashes with these values (for example) in the config.
```
"width" : 1999,
"height" : 1180,
```

rr- commented:
> I've noticed since previous checks the skybox appearing in the distance in e.g. Offshore Rig, but I'm wondering if this can be fixed by introducing draw dist/fade settings. And if so, it'd be fine as a follow up task IMO. I don't think this is a new issue in the PR by the way, I just skipped past certain levels with the problem.
> 
> ![image](https://private-user-images.githubusercontent.com/33758420/393646897-075d5cfc-0f4f-4a5c-826f-876f9b7abc20.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2OTY0NTgsIm5iZiI6MTczMzY5NjE1OCwicGF0aCI6Ii8zMzc1ODQyMC8zOTM2NDY4OTctMDc1ZDVjZmMtMGY0Zi00YTVjLTgyNmYtODc2ZjliN2FiYzIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA4VDIyMTU1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiZTUxNjdhYWU4YWIyMGVmYTExMWIzNThiZDRhODYyOTVmNzRiNmU5N2ZhZDI5NTY4NjFjYWM5YmZjNWVhYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nF2WuZ-4s4FCXyMEl55x9uLtwau46V6GIYXWsAl-5nE)
> 
> One bigger issue has appeared though since testing yesterday: SWR mode crashes for me in windowed mode, but it seems to be dependent on the window size. Unfortunately the log file doesn't produce a trace. The default window size is fine, but it consistently crashes with these values (for example) in the config.
> 
> ```
> "width" : 1999,
> "height" : 1180,
> ```

@lahm86 fixed the room clipping issue.

Regarding the crash, it'd be good to have anything that'd let us debug it.

aredfan commented:
- [x] I'm also able to reproduce the SWR crash described by Lahm. Attaching my TR2X.json5 where the game closes as soon as it opens unless I change the render mode from software to hardware within the json5 file. [TR2X.zip](https://github.com/user-attachments/files/18058212/TR2X.zip)

Switching back to SWR while in-game will cause the game to crash again. LMK if you need anything else.

____________________

- [x] If I use F1/Shift+F1 while in the stopwatch/end of level stats, this causes the UI to become persistent, and the game can crash too when the inputs are used several times.

____________________

- [ ] There's a graphical bug if I deselect the passport when in the Exit to Title/Exit Game pages. This issue might already be covered under the `ZBuffer near plane clipping` issue.

![20241209_074727_Laras_Home](https://github.com/user-attachments/assets/3f047a5d-69a7-4704-86f7-e61f6602a41c)

_________________

- [x] In Floating Islands, if I position the camera so that Lara's left arm is hidden from view, then release the look button, the game crashes.

**Save:** [savegame.zip](https://github.com/user-attachments/files/18060144/savegame.zip)

https://github.com/user-attachments/assets/8b92b12e-0ff8-4322-ba9b-38919c4ee6e4

___________________________

> /play 0 /tp 43 2 89 - depending on the setting, you will either hear the clock after pushing the button when the camera shifts or not.

When the microphone is set to Lara's position, I can still hear the ticking clock.

rr- committed:
tr2: use hybrid OpenGL/DirectX rendering

rr- committed:
tr2/rooms: fix crash when drawing too many rooms

Resolves #1998.

rr- committed:
tr2/requester: fix stats dialog cloning

Resolves #1999.

rr- commented:
> * [ ]  If I use F1/Shift+F1 while in the stopwatch/end of level stats, this causes the UI to become persistent, and the game can crash too when the inputs are used several times.

Thanks – it was an OG bug. I reported it as #1999 and have provided a fix.

> * [ ]  There's a graphical bug if I deselect the passport when in the Exit to Title/Exit Game pages. This issue might already be covered under the `ZBuffer near plane clipping` issue.

I can't reproduce it :( 

> * [ ]  In Floating Islands, if I position the camera so that Lara's left arm is hidden from view, then release the look button, the game crashes.
> 
> **Save:** [savegame.zip](https://github.com/user-attachments/files/18060144/savegame.zip)

It also turns out to be an OG bug that just became more noticeable after we fixed the skybox depth issues. I've tracked it as #1998 and have provided a fix. For thoroughness, since the skybox depth issue is also an OG bug, I've tracked that too as #2000.

> > /play 0 /tp 43 2 89 - depending on the setting, you will either hear the clock after pushing the button when the camera shifts or not.
> 
> * [ ]  When the microphone is set to Lara's position, I can still hear the ticking clock.

Thank you for noticing this, but I think it also happens in develop and since it's unrelated to rendering I'd like to keep it out of this PR. I think we already have a ticket for it and Lahm has even a fix pending.

aredfan commented:
Thank you for fixing many issues in this PR.  The test build is already looking very polished and stable.👍 

_______________________

- [x] I only found 1 new bug this time round. There's a stray shadow lurking in the Floating Islands skybox.

1. /play 16
2. hold the look button and gently move the camera downwards.

**Edit:** The stray shadow disappears when `/kill all` is used, it's probably an inactive enemy.
**Edit 2:** Item 43 - Guard (Sword) is the culpit.

![20241210_073620_Floating_Islands](https://github.com/user-attachments/assets/ca56a842-3f2c-4243-a27d-82a2b2305c0f)

______________________

> I can't reproduce it :(

The passport graphic bug only seems to happen when HWR and Z-buffer is enabled.

_____________________

One of the few remaining issues going unmentioned is the stretched pictures in 16:9/any aspect modes. TR1X properly scales the OG pictures when using widescreen.

![1](https://github.com/user-attachments/assets/966cd926-09d0-4949-999d-d6a4f6944908)

Btw I noticed that using widescreen aspect modes seem to cause cropping. TR1X has an option for Vertical FOV which fixes the cropping in widescreen.

![2](https://github.com/user-attachments/assets/0f579706-4861-46f7-8058-3103f8aca346)

I can raise the latter 2 issues separately if it's out of scope in this PR.

rr- commented:
Let's track these separately. I'll try to fix the shadows in this one.

walkawayy reviewed (approved):
LGTM

walkawayy commented:
Only thing I noticed was the z-buffer issues in HWR and SWR but it's under known issues.

![image](https://github.com/user-attachments/assets/9f3f01bd-db13-4d80-a6da-0e0b526b5b3e)

aredfan reviewed (approved):
A small issue I noticed when using shift+f7 - the OSD says `Z-Buffer On`.

Apart from that, everything LGTM, and thank you. 👍

rr- commented:
> Only thing I noticed was the z-buffer issues in HWR and SWR but it's under known issues.
> 
> ![image](https://private-user-images.githubusercontent.com/81546780/394504536-9f3f01bd-db13-4d80-a6da-0e0b526b5b3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM5MDcwNTcsIm5iZiI6MTczMzkwNjc1NywicGF0aCI6Ii84MTU0Njc4MC8zOTQ1MDQ1MzYtOWYzZjAxYmQtZGIxMy00ZDgwLWE2ZGEtMGUwYjUyNmI1YjNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjExVDA4NDU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBhZDMzNDJjNWFlNTVmMzIzNGJhZmJiY2RjY2Y1Njg5YmY0ZGQ5MDJhYzQ2ZGY1MGU0ZjgyN2EwNmZiOTJhN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Awmtx2VoRURYmKLLwTcFgIvXsDWRzusOkkhgnFGIXi8)

Thank you for re-testing. To clarify, I see the screenshot above is taken with the software renderer, so the z-buffer setting isn't relevant there, and this sorting issue matches the OG's behavior. The issues mainly arise in HWR with the Z buffer on, when we encounter depth races in shadows or culling of objects close to the camera, which differ from the original's behavior.

![20241211_085504_Laras_Home](https://github.com/user-attachments/assets/213616ce-4fc8-4d19-8c34-6c56204cc712)
![20241211_085503_Laras_Home](https://github.com/user-attachments/assets/b7abcda2-c32e-4da9-b043-f9b597ff463d)

rr- committed:
gfx: fix wireframe mode not working in tr2 title screen

rr- commented:
Rebased on top of #2004 which'll need to go first. Added a fix for the Z buffer near plane clipping, and an additional fix to support wireframe mode in the inventory screen.

rr- committed:
docs: update

rr- commented:
Added changes from #2007. This resolved visual depth issues when viewing multiple transparent surfaces that are either in close proximity or intersect with one another, with the z-buffer enabled.
@aredfan can I ask for one last round of sanity checks?

aredfan commented:
@rr- The transparent surfaces fix is working well, and it also fixes the same issue that was apparent in Lara's basement.

I also re-checked all the issues listed in the OP, LGTM.

walkawayy reviewed (commented):
LGTM!

walkawayy commented on docs/tr2/CHANGELOG.md:
```diff
@@ -1,38 +1,59 @@
+    - fixed black borders in windowed mode (#1645)
+    - fixed "Failed to create device" when toggling fullscreen (#1842)
+    - fixed distant rooms sometimes not appearing, causing the skybox to be visible when it shouldn't (#2000)
+- replaced fully the Windows Registry configuration with .json files
```
Nitpick but this is slightly awkward wording imo. I would remove `fully`.
